src/kernels/MIOpenTensorKernels.cl

```diff
@@ -395,102 +272,46 @@ __kernel void OpTensorLeadingOnesGeneric(global MIOPEN_TYPE* a,
     global MIOPEN_TYPE* a_off = a + Aoffset;
     global MIOPEN_TYPE* b_off = b + Boffset;
     global MIOPEN_TYPE* c_off = c + Coffset;
-#if FIRST_NOT_ONE == 3 // bitmap = 1,1,1,1
-    int tid = get_global_id(0);
-    // MIOPEN_TYPE operand = b[tid];
 
-    // num_wg: the number of workgroups should be launched
-    // MAX_NUM_WG: the maximum number of workgroups actually launched
-    for(; tid < num_wg; tid += MAX_NUM_WG)
-    {
-
-        int o_w = tid % c_w;
-        int o_h = (tid / c_w) % c_h;
-        int o_c = (tid / (c_w * c_h)) % c_c;
-        int o_n = tid / (c_w * c_h * c_c);
-
-        int aindex = o_n * a_nstride + o_c * a_cstride + o_h * a_hstride + o_w;
-        int bindex = o_n * b_nstride + o_c * b_cstride + o_h * b_hstride + o_w;
-        int cindex = o_n * c_nstride + o_c * c_cstride + o_h * c_hstride + o_w;
-        c_off[cindex] =
-            MIOPEN_TENSOR_OP(alpha0 * a_off[aindex], alpha1 * b_off[bindex]) + beta * c_off[cindex];
-    }
-
-#elif FIRST_NOT_ONE == 2 // bitmap = 1,1,1,0
-    int gid = get_group_id(0);
+    int gid = (bitmap == 15) ? get_global_id(0) : get_group_id(0);
 
     // num_wg: the number of workgroups should be launched
     // MAX_NUM_WG: the maximum number of workgroups actually launched
     for(; gid < num_wg; gid += MAX_NUM_WG)
     {
-        int lid = get_local_id(0);
-        int o_h = gid % c_h;
-        int o_c = (gid / c_h) % c_c;
-        int o_n = gid / (c_c * c_h);
-
-        int bindex          = o_n * b_nstride + o_c * b_cstride + o_h * b_hstride;
-        MIOPEN_TYPE operand = b_off[bindex] * alpha1;
-
-        while(lid < work_per_wg)
-        {
-            int aindex = o_n * a_nstride + o_c * a_cstride + o_h * a_hstride + lid;
-            int cindex = o_n * c_nstride + o_c * c_cstride + o_h * c_hstride + lid;
-            c_off[cindex] =
-                MIOPEN_TENSOR_OP(alpha0 * a_off[aindex], operand) + beta * c_off[cindex];
-
-            lid += get_local_size(0);
-        }
-    }
-#elif FIRST_NOT_ONE == 1 // bitmap = 1,1,0,0
-    int gid = get_group_id(0);
-
-    // num_wg: the number of workgroups should be launched
-    // MAX_NUM_WG: the maximum number of workgroups actually launched
-    for(; gid < num_wg; gid += MAX_NUM_WG)
-    {
-        int lid = get_local_id(0);
-        int o_c = gid % c_c;
-        int o_n = gid / c_c;
-
-        int bindex          = o_n * b_nstride + o_c * b_cstride;
+        int lid    = (bitmap == 15) ? 0 : get_local_id(0);
```

Comment:
Since we are checking bitmap at so many different variable initializations, we could reduce it one if/else block. Essentially, reducing condition check you are making for each variable.

e.g. 
```
if(bimap == 0x1)
{
    ow = gid % c_w;
    oh = c_w;
}
else if(bimap == 0x1)
{
   ow =  ?
   oh = ?
}

```

Reply:
if/else block is what I try to avoid because it has negative impact on performance, while ternary operator has less impact.

Reply:
It is a question of 1 if/else block vs quite a few ternary operators. In other words, consolidating a few conditional checks into a single one. I doubt this single #if else block will cause any thread divergence. Yes, it introduces conditional check but the same would be true for ternary operator.

Reply:
Try this if it works on our compiler

```
A conditional of the form “if-then-else” generates branching. Use the select() function to replace these structures with conditional assignments that do not cause branching. For example:

if(x==1) r=0.5;
if(x==2) r=1.0;

becomes

r = select(r, 0.5, x==1);
r = select(r, 1.0, x==2);

Note that if the body of the if statement contains an I/O, the if statement cannot be eliminated.
```

https://rocmdocs.amd.com/en/latest/Programming_Guides/Opencl-optimization.html?highlight=%27bool%27

Reply:
Some of the condition checks here are a little more complicated than simple if-then-else check. It may end up with several select function embedded in each other. The ternary should be as less expensive as the select function.

Reply:
Ok. I think we can improve on using ternary operator these many times. I would take your word for granted that either select or a single if/else wouldn't have made performance better.